docs/okeps/okep-5224-connecting-udns/okep-5224-connecting-udns.md

```diff
@@ -0,0 +1,938 @@
+------ --------------------------------------------------------------------------- --------
+pass   "(ip[4|6].src == <UDN_SUBNET> && ip[4|6].dst == <UDN_SUBNET>)"                1100
+drop   "(ip[4|6].src == $<ALL_CONNECTED_UDN_SUBNETS> && ip[4|6].dst == $<ALL_CONNECTED_UDN_SUBNETS>)"    1050
+```
+
+But these ACLs will be having `apply-after-lb=false` so the service
+DNAT will happen after the ACLs are applied. This guarantees traffic
+destined directly to the pods in other connected nteworks will be
+dropped but if its destined to a clusterIP service, then after the
+ACL stage, we will hit the load balancing stage in the pipeline
+where the DNAT to backend IP will happen. Remaining workflow is same
+as the Pods section where the traffic will be routed to the
+`vpnc-connect-router`.
+
+Idea2: If `PodNetworkConnectivity` is not set then we don't add
+any UDN subnet routes towards `vpnc-connect-router`. Instead, if
+`ClusterIPServiceNetworkConnectivity` is set, then we add routes
+for each clusterIP that belongs to the other connected networks
+to be routed to the `vpnc-connect-router`. We would need to then
+either host all load balancers for all connected networks on the
+`vpnc-connect-router` OR we would need to route the clusterIP to
+any random node-network pair belonging to the destination network
+where the DNAT and reroute would happen but that's complicated.
+
+Idea1 is easier. Will do a PoC to confirm it works.
+
+#### Overlapping Subnets connected via clusterIPs
+
+TBD - see if there are simpler ways via sNATing and masquerading we
+use for hairpin services that can be leveraged? Can I program the load
+balancer such that the masqueradeIP is set and hence in addition to DNAT
+we also SNAT to the origin network's masqueradeIP and that takes care of
+cross network traffic. Need to ensure the existing masqueradeIP is not
+used within/between networks.
+
+### Network Policies and Admin Network Policies
+
+If `PodNetworkConnectivity` is set, then Network Policies
+and AdminNetworkPolicies (when they get supported on UDNs)
+will that are created in the networks that are part of
+connect API will have peers that span across all connect networks.
+So the scope of policies will be within all connected networks.
+
+If `PodNetworkConnectivity` is not set, then policies won't
+apply across networks since there is no need to do so.
+
+### Controller Design / Changes to Components [WIP]
+
+**Cluster Manager**
+
+A new `vpnc-controller` will be created that watches the
+`VirtualPrivateNetworkConnect` objects and does the following:
+
+1) It is responsible to allocate a /31 range for each node-network pair:
+  * It get's the `connectSubnet` value from the VPNC and allocates
+    a `/31` point to point link subnet for each node-network pair
+  * These values are annotated on the `VirtualPrivateNetworkConnect`
+    CRD for it to be then consumed on the `ovnkube-controller` side.
+    Sample:
+    ```yaml
+    k8s.ovn.org/vpc-node-network-subnets: {
```

Comment:
can't we use node-id for this?

Reply:
hmm not sure I follow, we need a point-to-point link per node-network, did you mean somehow split for each node and then within each node use network-ID?
During our design call in June timeframe, we had decided its better to do this route since nodes and networks could dynamically come and go, so giving out /31 to each node-network pair is better. Instead of statically assigning it per node, that's also why I removed the need for hostsubnet under connect subnet in the API